internal/docstore/docstore.go

```diff
@@ -0,0 +1,216 @@
+// The document must not already exist; an AlreadyExists error is returned if it
+// does. If the document doesn't have key fields, it will be given key fields with
+// unique values.
+// TODO(jba): treat zero values for struct fields as not present?
+func (l *ActionList) Create(doc Document) *ActionList {
+	return l.add(&Action{kind: driver.Create, doc: doc})
+}
+
+// Replace adds an action that replaces a document.
+// The document must already exist; a NotFound error is returned if it does not.
+func (l *ActionList) Replace(doc Document) *ActionList {
+	return l.add(&Action{kind: driver.Replace, doc: doc})
+}
+
+// Put adds an action that adds or replaces a document.
+// The document may or may not already exist.
+func (l *ActionList) Put(doc Document) *ActionList {
+	return l.add(&Action{kind: driver.Put, doc: doc})
+}
+
+// Delete adds an action that deletes a document.
+// Only the key fields and RevisionField of doc are used.
+// If the document doesn't exist, nothing happens and no error is returned.
+func (l *ActionList) Delete(doc Document) *ActionList {
+	// Rationale for not returning an error if the document does not exist:
+	// Returning an error might be informative and could be ignored, but if the
+	// semantics of an action list are to stop at first error, then we might abort a
+	// list of Deletes just because one of the docs was not present, and that seems
+	// wrong, or at least something you'd want to turn off.
+	return l.add(&Action{kind: driver.Delete, doc: doc})
+}
+
+// Get adds an action that retrieves a document.
+// Only the key fields of doc are used to create the request.
+// If fps is omitted, all the fields of doc are set to those of the
+// retrieved document. If fps is present, only the given field paths are
+// set. In both cases, other fields of doc are not touched.
+//
+// A field path is a dot-separated sequence of UTF-8 field names. A field path
+// can select top level fields or elements of sub-documents. There is no way to
+// select a single list element.
+func (l *ActionList) Get(doc Document, fps ...string) *ActionList {
+	return l.add(&Action{
+		kind:       driver.Get,
+		doc:        doc,
+		fieldpaths: fps,
+	})
+}
+
+// Update applies Mods to doc, which must exist.
+// Only the key and revision fields of doc are used.
+//
+// A modification will create a field if it doesn't exist.
+//
+// No field path in mods can be a prefix of another. (It makes no sense
+// to, say, set foo but increment foo.bar.)
+//
+// Update does not modify its doc argument. To obtain the new value of the document,
+// call Get after calling Update.
+func (l *ActionList) Update(doc Document, mods Mods) *ActionList {
+	return l.add(&Action{
+		kind: driver.Update,
+		doc:  doc,
+		mods: mods,
+	})
+}
+
+// Mods is a map from field paths to modifications.
+// A field path is a dot-separated sequence of UTF-8 field names.
+// See ActionList.Update.
+type Mods map[string]interface{}
```

Comment:
Why a `map` and not a slice? Seems like you'd want this to be ordered....

Reply:
None of the document stores treats these as ordered.

The no-prefix rule (see `Update`), plus the fact that a mod cannot refer to a different field,  means that order doesn't matter.

It's also easier to write a literal map than a slice, because of the need for keyed struct literals. Compare
```
Mods{
    "a": 1,
    "b": 2,
}
```
with
```
Mods{
    {FieldPath: "a", Value: 1},
    {FieldPath: "b", Value: 2},
}
```